cmd/thanos/query.go

```diff
@@ -93,6 +93,9 @@ func registerQuery(m map[string]setupFunc, app *kingpin.Application, name string
 
 	storeResponseTimeout := modelDuration(cmd.Flag("store.response-timeout", "If a Store doesn't send any data in this specified duration then a Store will be ignored and partial data will be returned if it's enabled. 0 disables timeout.").Default("0ms"))
 
+	storeReadTimeout := modelDuration(cmd.Flag("store.read-timeout", "Maximum time to read response from store. If request to one of stores is timed out and store.read-timeout < query.timeout partial response will be returned. If store.read-timeout >= query.timeout one of stores is timed out the client will get no data and timeout error.").
```

Comment:
IMHO you are missing some `then`s and `a`s here. I would suggest to change it to:
```
Maximum time to read a response from a store. If a request to one of the stores has timed out and store.read-timeout < query.timeout then a partial response will be returned. If store.read-timeout >= query.timeout and one of the stores has timed out then the client will get no data, and the timeout error will be returned`.

Reply:
IMO this is the `query-timeout`, as it will time out the whole query if it doesn't succeed in +-10min. 

As the current `query-timeout` sometimes doesn't timeout the whole thing, so it should be `--promql-engine-timeout` or something like that.

@GiedriusS / @bwplotka wdyt?

Reply:
I agree :+1:

Reply:
Hey @povilasv @GiedriusS! This timeout is about particular store request, not for whole query.
From my point of view now `storeReadTimeout` and `storeResponseTimeout` looks a bit confusing and I would swap this parameters.

Reply:
So your timeout is:

If partial response is disabled, will timeout **the query** and won't return data.
If partial response is enabled, will timeout **the query** and will return data.

While store response timeout is:

Partial response is enabled: if store doesn't **respond** with any data in x seconds, query drops that Thanos Store API it and doesn't wait for it and returns partial response

If partial response is disabled then query just returns error.

So IMO **storeRepsonseTimeout** is about Thanos Store not responding in time.
Yours is about **query** not being able to finish in time.

Reply:
Yeah but this is implementation detail which users shouldn't have to think about, right? IMO implemenation wise It's a bit better place to put it in `proxy` layer instead of `query` layer, but the end effect is the same. The queries will be timed out in that time + small delta.

As all the `query` layer does is calls `proxy.Series` and returns response.
And at the begging of `proxy.Series` call we do:
```
	storeReadCtx, storeReadCancelFunc := s.contextWithReadTimeout(gctx)
	defer storeReadCancelFunc()
```

So whether we put it into `query` or at the begging of `proxy.Series` function the effect is the same.

Reply:
Cool,

re `store.response-timeout` -> `store.read-timeout`, we had a long conversation in that PR. We settled on response timeout for now, so let's put it off to a different PR if you want.

re `promql.engine-timeout` -> IMO good name, not a lot of people will understand, but we will set reasonable defaults so it shouldn't matter and for advanced users it might be important.

Reply:
okay, than only
`store.read-timeout` -> `promql.engine-timeout`

Reply:
So the timeout you introduced should be named `--store.read-timeout=10m` -> `--store.query-timeout`.

But because the `--store.query-timeout` is taken it needs to be renamed to `--promql.engine-timeout`

so:
1) `--store.query-timeout` -> `--promql.engine-timeout`
2) `--store.read-timeout` -> `--store.query-timeout`

Reply:
--store.query-timeout is not taken at the moment :)

Reply:
I mean this one

https://github.com/improbable-eng/thanos/pull/895/files#diff-fdba606ba45878a31172c82a2b017303L64

So then maybe:

`--query.timeout` -> `--promql-timeout`
`--store.read-timeout` -> `--query.timeout`

?
@bwplotka is our naming expert , wdyt?